## [E-02] 회고 링크
: https://github.com/zzozzo/AIFFEL/blob/master/EXPLORATION/%5BE02%5Ddigits_wine_canser/README.md

# 손글씨 분류하기

In [1]:
#모듈 import 하기
import pandas as pd

from sklearn import svm
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression


#데이터 불러오기
data = load_digits()
digit_data = data.data

In [2]:
#EDA

#Target Names 확인하기
print(data.target_names)

print()

print(data.feature_names)
#8*8사이즈 이미지의 각 픽셀

[0 1 2 3 4 5 6 7 8 9]

['pixel_0_0', 'pixel_0_1', 'pixel_0_2', 'pixel_0_3', 'pixel_0_4', 'pixel_0_5', 'pixel_0_6', 'pixel_0_7', 'pixel_1_0', 'pixel_1_1', 'pixel_1_2', 'pixel_1_3', 'pixel_1_4', 'pixel_1_5', 'pixel_1_6', 'pixel_1_7', 'pixel_2_0', 'pixel_2_1', 'pixel_2_2', 'pixel_2_3', 'pixel_2_4', 'pixel_2_5', 'pixel_2_6', 'pixel_2_7', 'pixel_3_0', 'pixel_3_1', 'pixel_3_2', 'pixel_3_3', 'pixel_3_4', 'pixel_3_5', 'pixel_3_6', 'pixel_3_7', 'pixel_4_0', 'pixel_4_1', 'pixel_4_2', 'pixel_4_3', 'pixel_4_4', 'pixel_4_5', 'pixel_4_6', 'pixel_4_7', 'pixel_5_0', 'pixel_5_1', 'pixel_5_2', 'pixel_5_3', 'pixel_5_4', 'pixel_5_5', 'pixel_5_6', 'pixel_5_7', 'pixel_6_0', 'pixel_6_1', 'pixel_6_2', 'pixel_6_3', 'pixel_6_4', 'pixel_6_5', 'pixel_6_6', 'pixel_6_7', 'pixel_7_0', 'pixel_7_1', 'pixel_7_2', 'pixel_7_3', 'pixel_7_4', 'pixel_7_5', 'pixel_7_6', 'pixel_7_7']


In [3]:
#ㅣabel 데이터 변수에 저장

digit_label = data.target
print(digit_label.shape)
digit_label

(1797,)


array([0, 1, 2, ..., 8, 9, 8])

In [4]:
print(data.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

In [5]:
#훈련과 테스트 셋 분류하기
X_train, X_test, y_train, y_test = train_test_split(digit_data, 
                                                    digit_label, 
                                                    test_size=0.2, 
                                                    random_state=7)

print('X_train 개수: ', len(X_train),', X_test 개수: ', len(X_test))

X_train 개수:  1437 , X_test 개수:  360


In [6]:
"""
학습시키기 #Decision Tree
"""

decision_tree = DecisionTreeClassifier(random_state=8)

decision_tree.fit(X_train, y_train)
y_pred_dt = decision_tree.predict(X_test)
accuracy = accuracy_score(y_test, y_pred_dt)
print("정확도: " ,accuracy)

print()
print()

"""
평가하기 #Decision Tree
"""
print("*****Decision Tree*****")
print(classification_report(y_test, y_pred_dt))
print()


정확도:  0.8444444444444444


*****Decision Tree*****
              precision    recall  f1-score   support

           0       0.98      0.98      0.98        43
           1       0.79      0.79      0.79        42
           2       0.82      0.82      0.82        40
           3       0.84      0.91      0.87        34
           4       0.79      0.92      0.85        37
           5       0.90      0.96      0.93        28
           6       0.93      0.93      0.93        28
           7       0.90      0.79      0.84        33
           8       0.80      0.65      0.72        43
           9       0.73      0.75      0.74        32

    accuracy                           0.84       360
   macro avg       0.85      0.85      0.85       360
weighted avg       0.85      0.84      0.84       360




In [7]:
"""
학습시키기 #Random Forest
"""

random_forest = RandomForestClassifier(random_state=32)

random_forest.fit(X_train, y_train)
y_pred_rf = random_forest.predict(X_test)
accuracy = accuracy_score(y_test, y_pred_rf)
print("정확도: " ,accuracy)

print()
print()

"""
평가하기 #Random Forest
"""
print("*****Random Forest*****")
print(classification_report(y_test, y_pred_rf))
print()

정확도:  0.9638888888888889


*****Random Forest*****
              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.93      1.00      0.97        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       0.93      1.00      0.96        37
           5       0.90      0.96      0.93        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       1.00      0.84      0.91        43
           9       0.94      0.94      0.94        32

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.97      0.96      0.96       360




In [8]:
"""
학습시키기 #SVM
"""

svm_clf = svm.SVC()

svm_clf.fit(X_train, y_train)
y_pred_svm = svm_clf.predict(X_test)
accuracy = accuracy_score(y_test, y_pred_svm)
print("정확도: " ,accuracy)

print()
print()

"""
평가하기 #SVM
"""
print("*****SVM*****")
print(classification_report(y_test, y_pred_svm))
print()

정확도:  0.9888888888888889


*****SVM*****
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      1.00      0.98        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        37
           5       0.93      1.00      0.97        28
           6       1.00      1.00      1.00        28
           7       1.00      1.00      1.00        33
           8       1.00      0.93      0.96        43
           9       1.00      0.97      0.98        32

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360




In [9]:
"""
학습시키기 #SGD Classifier
"""

sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred_sgd = sgd_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred_sgd)
print("정확도: " ,accuracy)

print()
print()

"""
평가하기 #SGD Classifier
"""
print("*****SGD Classifier*****")
print(classification_report(y_test, y_pred_sgd))
print()

정확도:  0.9361111111111111


*****SGD Classifier*****
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.78      0.95      0.86        42
           2       0.98      1.00      0.99        40
           3       0.97      0.91      0.94        34
           4       0.95      1.00      0.97        37
           5       0.96      0.93      0.95        28
           6       0.93      0.93      0.93        28
           7       0.97      0.97      0.97        33
           8       0.94      0.74      0.83        43
           9       0.94      0.94      0.94        32

    accuracy                           0.94       360
   macro avg       0.94      0.94      0.94       360
weighted avg       0.94      0.94      0.94       360




In [10]:
"""
학습시키기 #Logistic Regression
"""
logistic_model = LogisticRegression(max_iter = 4000)

logistic_model.fit(X_train, y_train)
y_pred_lr = logistic_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred_lr)
print("정확도: " ,accuracy)

print()
print()

"""
평가하기 #Logistic Regression
"""
print("*****Logistic Regression*****")
print(classification_report(y_test, y_pred_lr))
print()

정확도:  0.95


*****Logistic Regression*****
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      0.95      0.95        42
           2       0.98      1.00      0.99        40
           3       0.94      0.97      0.96        34
           4       1.00      1.00      1.00        37
           5       0.79      0.96      0.87        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       0.92      0.81      0.86        43
           9       0.97      0.88      0.92        32

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360




## 손글씨 분류 성능 지표 선택

In [11]:
#레이블별 데이터 수 확인

import collections
collections.Counter(digit_label)

Counter({0: 178,
         1: 182,
         2: 177,
         3: 183,
         4: 181,
         5: 182,
         6: 181,
         7: 179,
         8: 174,
         9: 180})

각 레이블별 데이터 수가 고르게 퍼져있음을 알 수 있다
따라서 *balanced data*라 볼 수 있기 때문에 **Accuracy**를 기준으로 모델 성능을 평가할 수 있다 판단하였다.

위 5가지 모델의 accuracy를 비교하였을 때 **SVM**의 성능이 **99%** 로 가장 높게 나와 SVM을 손글씨 분류 모델로 최종 선택한다.

# wine 분류기

In [51]:
from sklearn import svm
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

#데이터 불러오기
data = load_wine()
wine_data = data.data

In [52]:
#EDA

#Target Names 확인하기
print(data.target_names)

print()

print(data.feature_names)
#8*8사이즈 이미지의 각 픽셀

['class_0' 'class_1' 'class_2']

['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']


In [53]:
#ㅣabel 데이터 변수에 저장

wine_label = data.target
print(wine_label.shape)
wine_label

(178,)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2])

In [5]:
print(data.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

In [58]:
#훈련과 테스트 셋 분류하기
X_train, X_test, y_train, y_test = train_test_split(wine_data, 
                                                    wine_label, 
                                                    test_size=0.2, 
                                                    random_state=42)

print('X_train 개수: ', len(X_train),', X_test 개수: ', len(X_test))

X_train 개수:  142 , X_test 개수:  36


In [59]:
"""
학습시키기 #Decision Tree
"""

decision_tree = DecisionTreeClassifier(random_state=42)

decision_tree.fit(X_train, y_train)
y_pred_dt = decision_tree.predict(X_test)
accuracy = accuracy_score(y_test, y_pred_dt)
print("정확도: " ,accuracy)

print()
print()

"""
평가하기 #Decision Tree
"""
print("*****Decision Tree*****")
print(classification_report(y_test, y_pred_dt))
print()


정확도:  0.9444444444444444


*****Decision Tree*****
              precision    recall  f1-score   support

           0       0.93      0.93      0.93        14
           1       0.93      1.00      0.97        14
           2       1.00      0.88      0.93         8

    accuracy                           0.94        36
   macro avg       0.95      0.93      0.94        36
weighted avg       0.95      0.94      0.94        36




In [60]:
"""
학습시키기 #Random Forest
"""

random_forest = RandomForestClassifier(random_state=42)

random_forest.fit(X_train, y_train)
y_pred_rf = random_forest.predict(X_test)
accuracy = accuracy_score(y_test, y_pred_rf)
print("정확도: " ,accuracy)


print()
print()

"""
평가하기 #Random Forest
"""
print("*****Random Forest*****")
print(classification_report(y_test, y_pred_rf))
print()

정확도:  1.0


*****Random Forest*****
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        14
           1       1.00      1.00      1.00        14
           2       1.00      1.00      1.00         8

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36




In [61]:
"""
학습시키기 #SVM
"""

svm_clf = svm.SVC()

svm_clf.fit(X_train, y_train)
y_pred_svm = svm_clf.predict(X_test)
accuracy = accuracy_score(y_test, y_pred_svm)
print("정확도: " ,accuracy)

print()
print()

"""
평가하기 #SVM
"""
print("*****SVM*****")
print(classification_report(y_test, y_pred_svm))
print()

정확도:  0.8055555555555556


*****SVM*****
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        14
           1       0.73      0.79      0.76        14
           2       0.57      0.50      0.53         8

    accuracy                           0.81        36
   macro avg       0.77      0.76      0.76        36
weighted avg       0.80      0.81      0.80        36




In [63]:
"""
학습시키기 #SGD Classifier
"""

sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred_sgd = sgd_model.predict(X_test)
accuracy = accuracy_score(y_test1, y_pred_sgd)
print("정확도: " ,accuracy)

print()
print()

"""
평가하기 #SGD Classifier
"""

print("*****SGD Classifier*****")
print(classification_report(y_test, y_pred_sgd1,zero_division=1))
print()

정확도:  0.75


*****SGD Classifier*****
              precision    recall  f1-score   support

           0       1.00      0.21      0.35        14
           1       0.42      1.00      0.60        14
           2       1.00      0.00      0.00         8

    accuracy                           0.47        36
   macro avg       0.81      0.40      0.32        36
weighted avg       0.78      0.47      0.37        36




In [70]:
"""
학습시키기 #Logistic Regression
"""
logistic_model = LogisticRegression(max_iter = 3000)

logistic_model.fit(X_train, y_train)
y_pred_lr = logistic_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred_lr)
print("정확도: " ,accuracy)

print()
print()

"""
평가하기 #Logistic Regression
"""
print("*****Logistic Regression*****")
print(classification_report(y_test, y_pred_lr))
print()

정확도:  1.0


*****Logistic Regression*****
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        14
           1       1.00      1.00      1.00        14
           2       1.00      1.00      1.00         8

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36




## 와인 분류 성능 지표 선택

In [22]:
#레이블별 데이터 수 확인

import collections
collections.Counter(wine_label)

Counter({0: 59, 1: 71, 2: 48})

1번과 2번 레이블의 데이터 차이가 비교적 많이 나기 때문에 Accuracy만으로 평가하기에 부족하다 생각하였다. 따라서 precision을 추가로 고려하도록 한다.

In [71]:
#각 모델 precision확인
from sklearn.metrics import precision_score

print("--- Decision Tree precision ---")
print(precision_score(y_test, y_pred_dt,average='micro'))
print()
print("--- Random Forest precision ---")
print(precision_score(y_test,y_pred_rf,average='micro'))
print()
print("--- SVM precision ---")
print(precision_score(y_test, y_pred_svm,average='micro'))
print()
print("--- SGD Classifier precision ---")
print(precision_score(y_test, y_pred_sgd,average='micro'))
print()
print("--- Logistic Regression precision ---")
print(precision_score(y_test, y_pred_lr,average='micro'))
print()

--- Decision Tree precision ---
0.9444444444444444

--- Random Forest precision ---
1.0

--- SVM precision ---
0.8055555555555556

--- SGD Classifier precision ---
0.75

--- Logistic Regression precision ---
1.0



위 5가지 모델의 **precision**을 비교하였을 때 **Random Forest**와 **Logistic Regression**의 성능이 **100%** 로 가장 높게 나왔다. 하지만 Logistic Regression의 경우 *max_iter 인자를 추가하기 전 약 98%의 pesicion의 값*이 나왔기 때문에, **Random Forest를 기반**으로 하되, 상황에 따라 Logistic Regression을 커스텀 하여 사용하면 될 것 같다.

# 유방암 예측하기

In [41]:
from sklearn import svm
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

data = load_breast_cancer()
cancer_data = data.data

In [42]:
#EDA

#Target Names 확인하기
print(data.target_names)

print()

print(data.feature_names)
#8*8사이즈 이미지의 각 픽셀

['malignant' 'benign']

['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']


In [43]:
#ㅣabel 데이터 변수에 저장

cancer_label = data.target
print(cancer_label.shape)
cancer_label

(569,)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

In [26]:
print(data.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

In [44]:
#훈련과 테스트 셋 분류하기
X_train, X_test, y_train, y_test = train_test_split(cancer_data, 
                                                    cancer_label, 
                                                    test_size=0.2, 
                                                    random_state=42)

print('X_train 개수: ', len(X_train),', X_test 개수: ', len(X_test))

X_train 개수:  455 , X_test 개수:  114


In [45]:
"""
학습시키기 #Decision Tree
"""

decision_tree = DecisionTreeClassifier(random_state=8)

decision_tree.fit(X_train, y_train)
y_pred_dt = decision_tree.predict(X_test)
accuracy = accuracy_score(y_test, y_pred_dt)
print("정확도: " ,accuracy)

print()
print()

"""
평가하기 #Decision Tree
"""
print("*****Decision Tree*****")
print(classification_report(y_test, y_pred_dt))
print()


정확도:  0.9298245614035088


*****Decision Tree*****
              precision    recall  f1-score   support

           0       0.91      0.91      0.91        43
           1       0.94      0.94      0.94        71

    accuracy                           0.93       114
   macro avg       0.93      0.93      0.93       114
weighted avg       0.93      0.93      0.93       114




In [46]:
"""
학습시키기 #Random Forest
"""

random_forest = RandomForestClassifier(random_state=42)

random_forest.fit(X_train, y_train)
y_pred_rf = random_forest.predict(X_test)
accuracy = accuracy_score(y_test, y_pred_rf)
print("정확도: " ,accuracy)


print()
print()

"""
평가하기 #Random Forest
"""
print("*****Random Forest*****")
print(classification_report(y_test, y_pred_rf))
print()

정확도:  0.9649122807017544


*****Random Forest*****
              precision    recall  f1-score   support

           0       0.98      0.93      0.95        43
           1       0.96      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.97      0.96      0.96       114
weighted avg       0.97      0.96      0.96       114




In [47]:
"""
학습시키기 #SVM
"""

svm_clf = svm.SVC()

svm_clf.fit(X_train, y_train)
y_pred_svm = svm_clf.predict(X_test)
accuracy = accuracy_score(y_test, y_pred_svm)
print("정확도: " ,accuracy)

print()
print()

"""
평가하기 #SVM
"""
print("*****SVM*****")
print(classification_report(y_test, y_pred_svm))
print()

정확도:  0.9473684210526315


*****SVM*****
              precision    recall  f1-score   support

           0       1.00      0.86      0.92        43
           1       0.92      1.00      0.96        71

    accuracy                           0.95       114
   macro avg       0.96      0.93      0.94       114
weighted avg       0.95      0.95      0.95       114




In [48]:
"""
학습시키기 #SGD Classifier
"""

sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred_sgd = sgd_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred_sgd)
print("정확도: " ,accuracy)

print()
print()

"""
평가하기 #SGD Classifier
"""

print("*****SGD Classifier*****")
print(classification_report(y_test, y_pred_sgd))
print()

정확도:  0.868421052631579


*****SGD Classifier*****
              precision    recall  f1-score   support

           0       0.74      1.00      0.85        43
           1       1.00      0.79      0.88        71

    accuracy                           0.87       114
   macro avg       0.87      0.89      0.87       114
weighted avg       0.90      0.87      0.87       114




In [49]:
"""
학습시키기 #Logistic Regression
"""
logistic_model = LogisticRegression(max_iter = 4000)

logistic_model.fit(X_train, y_train)
y_pred_lr = logistic_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred_lr)
print("정확도: " ,accuracy)

print()
print()

"""
평가하기 #Logistic Regression
"""
print("*****Logistic Regression*****")
print(classification_report(y_test, y_pred_lr))
print()

정확도:  0.956140350877193


*****Logistic Regression*****
              precision    recall  f1-score   support

           0       0.97      0.91      0.94        43
           1       0.95      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114




## 유방암 여부 분류 성능 지표 선택

In [50]:
#각 모델 recall확인
from sklearn.metrics import recall_score

print("--- Decision Tree Recall ---")
print(recall_score(y_test, y_pred_dt,average='micro'))
print()
print("--- Random Forest Recall ---")
print(recall_score(y_test,y_pred_rf,average='micro'))
print()
print("--- SVM Recall ---")
print(recall_score(y_test, y_pred_svm,average='micro'))
print()
print("--- SGD Classifier Recall ---")
print(recall_score(y_test, y_pred_sgd,average='micro'))
print()
print("--- Logistic Regression Recall ---")
print(recall_score(y_test, y_pred_lr,average='micro'))
print()

--- Decision Tree Recall ---
0.9298245614035088

--- Random Forest Recall ---
0.9649122807017544

--- SVM Recall ---
0.9473684210526315

--- SGD Classifier Recall ---
0.868421052631579

--- Logistic Regression Recall ---
0.956140350877193



유방암 판단 여부는 암인데 암이 아닌 것으로 판단하는 비율을 줄여야 하기 때문에 recall을 평가 지표로 하기로 했다.
! recall : 암이 아닌 것으로 판단하는 경우가 적어야 할 때

위 5가지 모델의 recall을 비교하였을 때 Random Forest와 Logistic Regression의 순으로 Recall 값이 가장 높은 것을 확인할 수 있다. 두 모델의 accuracy는 모두 약 96%이며 recall의 값도 약 96%이므로 두 모델을 함께 사용해도 좋을 것 같다.